### ISSUE TITLE
Output params file

### ISSUE THREAD

**Issue by JoseEspinosa** [nextflow-io/nextflow#1515]
## Output params file

Add an option in the command line allowing to keep all the parameters used into a `yaml` or `json` file.

## Usage scenario 

It would be useful to keep the parameters used in a given run in a systematic manner. This file containing the parameters can be then provided in a new execution to reproduce the exact command line. 

## Suggest implementation 

This might be implemented in `CmdRun.groovy`

It should allow choosing between the file formats `yaml` and `json`. Something like `params-file-yaml` or `params-file-json`

**Comment by pditommaso** [comment]
Not sure to understand the use case here. The user is expected to use the nextflow config/profile to have a consistent replicable configuration. You can use `nextflow config <pipeline>` to inspect the actual config.

**Comment by JoseEspinosa** [comment]
The idea here was to have the counterpart of the CLI option `-params-file`.

**Comment by pditommaso** [comment]
Which is the use case? for provenance and documentation purpose? In that case it would be better to include the params in the html report created by NF.

**Comment by JoseEspinosa** [comment]
The use case will be to obtain a file with the exact parameters used in a run in a `yaml` or a `json` file. This file can be then used to run the pipeline using the same parametrization by providing it with  `-params-file` option.

**Comment by pditommaso** [comment]
Still thinking it will be a redundant option. Users can just create a yml file on the first place instead of typing the params on the command line. IMO would be useful to include this info in the html report.

**Comment by ewels** [comment]
x-ref (basically duplicates): https://github.com/nextflow-io/nextflow/issues/1682 and https://github.com/nf-core/tools/issues/747

Use case would be some kind of provenance reporting, alongside the pipeline reports. Nextflow can take in configs from multiple layers and multiple locations, so it's nice for future reproducibility to be able to see the complete results of the config used for a pipeline. Then in theory the pipeline could be re-run with that file as `-params-file` to get the same output, irrespective of whether the user had a secret `~/.nextflow/config` file lurking around with a bunch of stuff they didn't realise that they had.

Bonus feature would be to be able to dump this into MultiQC (even as a proper module if we have a standardised format / filename etc). But this part could be complicated if doing it as a pipeline report type thing as those typically aren't available within a pipeline run.

**Comment by pditommaso** [comment]
Hey Phil, I agree that would useful to add the resolved config and params into the execution report. Less convinced to produce these as yet another file, that in the most case would get lost.

**Comment by JoseEspinosa** [comment]
I will argue that having this information on the execution report, although useful, won't allow its programmatic access. Dumping everything in a `yml` or `json` file will enable reproducing the same execution just using the `-params-file` option, which should be the main benefit of this feature request. It will be kind of a mirror option for `params-file`, if this make sense.

**Comment by ewels** [comment]
Can have both? 😉 If off by default then I don't see the file clutter being a problem (though we would probably have it on by default for @nf-core pipelines).

**Comment by ewels** [comment]
Brought up again in Euro-FAANG meeting, still definitely interest in having this as a feature..

**Comment by ewels** [comment]
@JoseEspinosa remind me - is there a reason that we can't do this within the pipeline script rather than having it as a core Nextflow feature? eg. Dumping `params` to a file. Could potentially just add it to the nf-core template then.

**Comment by JoseEspinosa** [comment]
If I remember well there is no reason to do this within the pipeline script. Only, that to me it makes more sense to have it at the Nextflow level for the reasons already discussed on this thread:
> I will argue that having this information on the execution report, although useful, won't allow its programmatic access. Dumping everything in a `yml` or `json` file will enable reproducing the same execution just using the `-params-file` option, which should be the main benefit of this feature request. It will be kind of a mirror option for `params-file`, if this makes sense.

**Comment by stevekm** [comment]
Definitely interested in this as a feature, in fact its become a critical requirement for us in some cases, especially when there are dynamic params and configs getting resolved inside the pipeline, in order to know after the run completes (and for record keeping) the exact params and configs that were used. This is especially important with things like dynamic process directives. 

Currently, I am simply doing this inside my pipeline;

```groovy

import groovy.json.JsonOutput

process SAVE_PARAMS {
    input:
    tuple val(workflowParams), val(outputFilename)

    output:
    path(outputFilename)

    script:
    json_str = JsonOutput.toJson(workflowParams)
    json_indented = JsonOutput.prettyPrint(json_str)
    // NOTE: single quotes are critical here;
    """
    echo '${json_indented}' > "${outputFilename}"
    """
}

SAVE_PARAMS(params, "params.json")
```

This is a pretty janky method, because it requires using `echo` inside a `script` block to execute (via the shell) to save a blob of unknown JSON text to file.
I also often get warning messages about the `script` block contents being too large when doing this too. 

It would be nice if I could at least use `exec` to do this directly in Groovy, but [that is not an option](https://github.com/nextflow-io/nextflow/issues/3962)

**Comment by ewels** [comment]
@stevekm - you can use the [`collectFile`](https://www.nextflow.io/docs/latest/operator.html#collectfile) channel operator to convert a value into a file. That's what we do [in the nf-core template](https://github.com/nf-core/tools/blob/e5ce6ce20304835bd40f102f038b7e1aadc888b2/nf_core/pipeline-template/workflows/pipeline.nf#L99-L100) to take pass workflow parameters to MultiQC. 

We also use native Groovy [to write files outside of `process` context](https://github.com/nf-core/tools/blob/e5ce6ce20304835bd40f102f038b7e1aadc888b2/nf_core/pipeline-template/lib/NfcoreTemplate.groovy#L156-L164) for the summary HTML + txt. I'm not sure if this is necessarily a good idea though and @pditommaso might cry a little if he realises that we're doing this 😆  But it's been in the nf-core template for years and I don't remember us getting many complaints, so it's probably mostly working..?

I agree that this would be a nice core feature to enable in the same way as we do for trace files etc (`trace.enabled` - so could be `resolved_params.enabled` or whatever). Should be minimal code required and will not impact any users not specifically looking for it.

**Comment by bentsherman** [comment]
I think the `nextflow config` command could do exactly this thing for you, if we added the same CLI options for params and params file. There is an option now to query a particular setting, we could extend that to support entire scopes, so you could say `-value params` and get all the params. Maybe also support other output formats like JSON / YAML, although that's starting to go beyond the scope of the `config` command.

**Comment by ewels** [comment]
In this case I think we want is as a `run` standard output though, rather than in a separate command. The idea being that it's partly provenance / future re-run and so you don't really know when it'll be needed.

**Comment by bentsherman** [comment]
Well, we could simplify @stevekm 's example to a regular Groovy function, or use the workflow complete handler:

```groovy
import groovy.json.JsonOutput

workflow.onComplete {
    jsonStr = JsonOutput.toJson(params)
    file("${params.outdir}/params.json").text = JsonOutput.prettyPrint(jsonStr)
}
```

**Comment by ewels** [comment]
Yup, let's do this in the nf-core template then. @JoseEspinosa do you fancy making an issue / PR to the template in nf-core/tools with @bentsherman's snippet? 🙏🏻 

I still think that this would be a good core feature for Nextflow, but we can get it in all nf-core pipelines pretty fast so let's do that in the short term.

Thread TLDR; proposed implementation for core Nextflow, long term solution:

```nextflow
// Showing suggested default values
paramsExport {
    enabled = false
    file = "params-<timestamp>.yml"
    overwrite = false
}
```

**Comment by bentsherman** [comment]
@stevekm the params are included in the BCO manifest
